### PR TITLE
solana transfer now requires --allow-unfunded-recipient if the recipient doesn't exist

### PR DESCRIPTION
**Pull request by mvines**
#### Problem
Users are losing funds because we never implemented https://github.com/solana-labs/solana/issues/6970

#### Summary of Changes
```
$ solana-keygen new -o ./new-account.json
$ solana transfer ./new-account.json 1
Error: The recipient address (9WtBoGqU1PwAvZwRGiS17zEhr8srDw8KdbSMxHkBpQuk) is not funded. Add `--allow-unfunded-recipient` to complete the transfer 
$ solana transfer ./new-account.json 1 --allow-unfunded-recipient
Signature: 4Z8xMnFqQKmHepZg6cotwHXdKJyTAtgv7DdKGWK4Y9aTAfJKTZku1RE56VWH41fx6enwVDaZQ6tgcMU8REmdPzVu
```

This now aligns with how `spl-token transfer` works as well.

#### Migration plan:
* Backport the `--allow-unfunded-recipient` flag to v1.5 as a hidden/ignored argument (https://github.com/solana-labs/solana/pull/16060)
* Update Exchange docs to suggest `--allow-unfunded-recipient` be now used, and update known users
* Backport this full PR to v1.6

This is a breaking change to the CLI, but a loss of funds is worse.

### CODE REVIEW

**Comment by t-nelson** [comment]
Should we frame this in terms of the account not existing rather than by balance? The notion of "empty" doesn't seem to communicate much w/o context. "Recipient does not exist, pass --allow-create-recipient" on the other hand is quite clear

**Comment by mvines** [comment]
> Should we frame this in terms of the account not existing rather than by balance? The notion of "empty" doesn't seem to communicate much w/o context. "Recipient does not exist, pass --allow-create-recipient" on the other hand is quite clear

The recipient likely does exist though, that is the keypair exists.  There's just no on-chain balance held by the account.

I feel like saying recipient doesn't exist would be more confusing.  "It does exist!  I `solana-keygen new`ed it"

**Comment by t-nelson** [comment]
> The recipient likely does exist though, that is the keypair exists. There's just no on-chain balance held by the account.
> 
> I feel like saying recipient doesn't exist would be more confusing. "It does exist! I `solana-keygen new`ed it"

Hmm... I've been trying to correct this notion of "my keypair is my account" whenever I come across it (usually new validators who botched vote account creation).  Do you think this is a wide-spread view of how things work?

Regardless, my main gripe is with "empty", I would probably allow something more explicit "zero-balance"

**Comment by mvines** [comment]
> Regardless, my main gripe is with "empty", I would probably allow something more explicit "zero-balance"

`--alllow-unfunded-recipient`?

Here's what sollet presents in this case:
image

**Comment by t-nelson** [comment]
> `--allow-unfunded-recipient`?
> 
> Here's what sollet presents in this case:
> image
> 
> or maybe just: `--unfunded-recipient`

I can get on board with that!

Part of me wants to suggest that it's safe to suppress this warning in the case of valid 44-char base58 encoded pubkeys, but someone mentioned recently that a user didn't realize the pubkeys were case-sensitive and burned funds that way :facepalm:

**Comment by mvines** [comment]
> Part of me wants to suggest that it's safe to suppress this warning in the case of valid 44-char base58 encoded pubkeys, but someone mentioned recently that a user didn't realize the pubkeys were case-sensitive and burned funds that way 🤦

Not safe!  Copy-and-paste an address and missing the first or last character and you likely have a valid 43-char base58 encoded pubkey.  😭

**Comment by t-nelson** [comment]
> ...  valid 44-char ...

But yeah, lots of ways to screw this up.  Once we invent time travel I'm going back to push harder for checksums, breakage be damned

File: cli/tests/transfer.rs [hunk 1]
```diff
@@ -423,6 +434,53 @@ fn test_transfer_all() {
...
+        blockhash_query: BlockhashQuery::All(blockhash_query::Source::Cluster),
+        nonce_account: None,
+        nonce_authority: 0,
+        fee_payer: 0,
+        derived_address_seed: None,
+        derived_address_program_id: None,
+    };
+
+    // Expect failure due to unfunded recipient and the lack of the `allow_unfunded_recipient` flag
+    process_command(&config).unwrap_err();
```
Comment (mvines) [inline cli/tests/transfer.rs:481]: new test

File: cli/src/cli.rs [hunk 2]
```diff
@@ -2945,11 +2974,12 @@ mod tests {
             }
         );
 
-        // Test Transfer no-wait
+        // Test Transfer no-wait and --allow-unfunded-recipient
         let test_transfer = test_commands.clone().get_matches_from(vec![
             "test",
             "transfer",
             "--no-wait",
+            "--allow-unfunded-recipient",
```
Comment (mvines) [inline cli/src/cli.rs:2982]: new test

**Comment by mvines** [comment]
@ryoqun - heads up that this'll affect at least one exchange

**Comment by mvines** [comment]
Travis is red due to:
> anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

Ugh

**Comment by mvines** [comment]
@t-nelson - does this seem ok?  You r+'ed the other support PRs but not the most important one, this one :)

**Comment by t-nelson** [comment]
> @t-nelson - does this seem ok? You r+'ed the other support PRs but not the most important one, this one :)

Whoops sorry! I thought I already reviewed this one. Sec..

**Review by t-nelson — DISMISSED** [review]
lgtm with the suggestion applied (feel free to DeMorgan)

File: cli/src/cli.rs [hunk 3]
```diff
@@ -1153,6 +1157,21 @@ fn process_transfer(
     let (recent_blockhash, fee_calculator) =
         blockhash_query.get_blockhash_and_fee_calculator(rpc_client, config.commitment)?;
 
+    if !allow_unfunded_recipient {
```
[the remark this answers is not in the document]
Comment (mvines) [reply cli/src/cli.rs:1160]: ah, cool

**Pull request merged**